cap/error.go

```diff
@@ -3,169 +3,159 @@ package cap
 import (
 	"errors"
 	"fmt"
+	"sort"
 	"strings"
 
 	"github.com/capatazlib/go-capataz/internal/c"
 )
 
-// startError is the error reported back to a Supervisor when the start of a
-// worker fails
-type startError = error
+type terminateNodeError = error
 
-// terminateError is the error reported back to a Supervisor when the
-// termination of a worker fails
-type terminateError = error
+type startNodeError = error
 
-// SupervisorError wraps an error from a supervised
-// worker, enhancing it with supervisor information and possible shutdown errors
-// on other siblings
-type SupervisorError struct {
+// ErrKVs is an utility interface used to get key-values out of Capataz errors
+type ErrKVs interface {
+	KVs() map[string]interface{}
+}
+
+// SupervisorTerminationError wraps errors returned by a child node that failed
+// to terminate (io errors, timeouts, etc.), enhancing it with supervisor
+// information
+type SupervisorTerminationError struct {
 	supRuntimeName string
-	rscCleanupErr  error
-	nodeErr        error
 	nodeErrMap     map[string]error
+	rscCleanupErr  error
 }
 
-// Unwrap returns anj error from a supervised goroutine (if any)
-func (se *SupervisorError) Unwrap() error {
-	return se.nodeErr
+// Error returns an error message
+func (err *SupervisorTerminationError) Error() string {
+	return "supervisor terminated with failures"
 }
 
-// Cause returns an error from a supervised goroutine (if any)
-func (se *SupervisorError) Cause() error {
-	return se.nodeErr
+// KVs returns a metadata map for structured logging
+func (err *SupervisorTerminationError) KVs() map[string]interface{} {
+	nodeNames := make([]string, 0, len(err.nodeErrMap))
+	for nodeName := range err.nodeErrMap {
+		nodeNames = append(nodeNames, nodeName)
+	}
+	sort.Strings(nodeNames)
+
+	acc := make(map[string]interface{})
+	acc["supervisor.name"] = err.supRuntimeName
+
+	for i, nodeName := range nodeNames {
+		nodeErr := err.nodeErrMap[nodeName]
+		var subTreeError ErrKVs
+		if errors.As(nodeErr, &subTreeError) {
+			for k0, v := range subTreeError.KVs() {
+				k := strings.TrimPrefix(k0, "supervisor.")
+				acc[fmt.Sprintf("supervisor.subtree.%d.%s", i, k)] = v
+			}
+		} else {
+			acc[fmt.Sprintf("supervisor.termination.node.%d.name", i)] = nodeName
+			acc[fmt.Sprintf("supervisor.termination.node.%d.error", i)] = nodeErr
```

Comment:
Something does not feel right here. Instead of `interface{}` why not use a structure with a `name` and an `err` in it, along with and appropriate `String()` and just have one entry? I would just add a `String()` to `SupervisorTerminationError` and put it in the `KV`. To me `interface{}` is just a source of mystery and confusion. I would suggest being explicit.

Reply:
I think some context is needed as to why this is `interface{}`. The goal is to do `errors.New("doge error").KVs(errors.F(err.KVs()))` and `log.KVs(log.F(err.KVs())` whenever we get an error from Capataz. To do this, we must return a `map[string]interface{}` on the `KVs` function.

Reply:
My brain read these as individual logs. I was completely upside down. So sorry. One comment I would make, from looking at other structured logs in kibana, is that the keys are getting really long and overwhelm other content (namely the actual values). Adding them as columns becomes difficult. This is just a subjective observation. Consider `killed` or `stop` in place of termination, for example. Eliminate `node`, unless there is something else that can be terminated? Maybe `cap` instead of `supervisor`, or just `sup`. Just a thought.